sno/merge.py

```diff
@@ -34,66 +34,164 @@ def merge(ctx, ff, ff_only, commit):
+    c_ours = repo[repo.head.target]
 
     # accept ref-ish things (refspec, branch, commit)
-    c_head, r_head = repo.resolve_refish(commit)
+    c_theirs, r_theirs = repo.resolve_refish(commit)
 
-    print(f"Merging {c_head.id} to {c_base.id} ...")
-    merge_base = repo.merge_base(c_base.oid, c_head.oid)
-    print(f"Found merge base: {merge_base}")
+    print(f"Merging {c_theirs.id} to {c_ours.id} ...")
+    merge_base_id = repo.merge_base(c_theirs.id, c_ours.id)
+    print(f"Found merge base: {merge_base_id}")
+
+    if not merge_base_id:
+        raise InvalidOperation(f"Commits {c_theirs.id} and {c_ours.id} aren't related.")
 
     # We're up-to-date if we're trying to merge our own common ancestor.
-    if merge_base == c_head.oid:
+    if merge_base_id == c_theirs.id:
         print("Already merged!")
         return
 
     # We're fastforwardable if we're our own common ancestor.
-    can_ff = merge_base == c_base.id
+    can_ff = merge_base_id == c_ours.id
 
     if ff_only and not can_ff:
         print("Can't resolve as a fast-forward merge and --ff-only specified")
         ctx.exit(1)
 
     if can_ff and ff:
         # do fast-forward merge
-        repo.head.set_target(c_head.id, "merge: Fast-forward")
-        commit_id = c_head.id
+        repo.head.set_target(c_theirs.id, "merge: Fast-forward")
+        commit_id = c_theirs.id
         print("Fast-forward")
     else:
-        ancestor_tree = repo[merge_base].tree
-
+        c_ancestor = repo[merge_base_id]
         merge_index = repo.merge_trees(
-            ancestor=ancestor_tree, ours=c_base.tree, theirs=c_head.tree
+            ancestor=c_ancestor.tree, ours=c_ours.tree, theirs=c_theirs.tree
         )
         if merge_index.conflicts:
-            print("Merge conflicts!")
-            repo_structure = RepositoryStructure(repo)
-            for (ancestor, ours, theirs) in merge_index.conflicts:
-                dataset = repo_structure.get_for_index_entry(ours)
-                pk = dataset.index_entry_to_pk(ours)
-                print(f"Conflict: {dataset.path}:{pk}")
-            raise NotYetImplemented("Sorry, merging conflicts isn't supported yet")
-
-        print("No conflicts!")
-        merge_tree_id = merge_index.write_tree(repo)
-        print(f"Merge tree: {merge_tree_id}")
-
-        user = repo.default_signature
-        merge_message = "Merge '{}'".format(r_head.shorthand if r_head else c_head.id)
-        commit_id = repo.create_commit(
-            repo.head.name,
-            user,
-            user,
-            merge_message,
-            merge_tree_id,
-            [c_base.oid, c_head.oid],
-        )
-        print(f"Merge commit: {commit_id}")
+            handle_merge_conflicts(
+                repo,
+                merge_index,
+                ancestor=(c_ancestor, "ancestor"),
+                ours=(c_ours, "HEAD"),
+                theirs=(c_theirs, commit),
+            )
+        else:
+            print("No conflicts!")
```

Comment:
Should probably make this output the commit info (either the merge commit or the HEAD after fast-forward), as JSON/text/etc?

(separate PR is fine, as is cleaning this up after merge handling is completed)

Reply:
And I guess logging/progress/etc should go to stderr.